### PR TITLE
protobuf: 3.0.0-alpha-3.1 -> 3.0.0-beta-2

### PR DESCRIPTION
**Pull request by eduarrrd**
Also split out gmock's source so that it can be copied into protobuf's
source. Hopefully this hack can be removed again once gmock is
replaced by gtest.

Also introduce my company to the COPYING doc since they own the
copyright on my contributions.

### CODE REVIEW

File: pkgs/development/libraries/protobuf/3.0.nix [hunk 1]
```diff
@@ -1,12 +1,46 @@
-{ callPackage, fetchFromGitHub, ... } @ args:
+{ stdenv, fetchFromGitHub
+, autoreconfHook, zlib, gmock
+, ...
```
Comment (edolstra) [inline pkgs/development/libraries/protobuf/3.0.nix]: Why `...`?

Comment (eduarrrd) [inline pkgs/development/libraries/protobuf/3.0.nix]: Good question. I missed this while merging. Let me remove that.

Comment (eduarrrd) [inline pkgs/development/libraries/protobuf/3.0.nix]: Fixed.

File: COPYING [hunk 2]
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2003-2006 Eelco Dolstra
+Copyright (c) 2003-2006 Eelco Dolstra, Google Inc.
```
Comment (pSub) [inline COPYING]: Why Google Inc.?

Comment (eduarrrd) [inline COPYING]: Because according my terms of employment say that they own the copyright on anything I produce while I work for them.

If you question is why I'm including them in this file? Because that's part of the terms for me (=them) to contribute here.

What it means? It just naming one (out of many) other copyright holders but Eelco. (Unless there's a CLA that reassigns all copyright to him that I missed.)

Comment (jagajaga) [inline COPYING]: I'm totally against that line.

File: pkgs/development/libraries/protobuf/3.0.nix [hunk 3]
```diff
@@ -1,12 +1,43 @@
...
+    substituteInPlace src/google/protobuf/testing/googletest.cc \
+      --replace 'tmpnam(b)' '"'$TMPDIR'/foo"'
+  '';
+
+  buildInputs = [ autoreconfHook zlib ];
+
+  doCheck = true;
+
+  meta = {
+    description = "Protocol Buffers - Google's data interchange format";
```
Comment (jagajaga) [inline pkgs/development/libraries/protobuf/3.0.nix]: Omit pkg's name in the description.

File: pkgs/development/libraries/protobuf/3.0.nix [hunk 4]
```diff
@@ -1,12 +1,43 @@
...
+
+    longDescription =
+      '' Protocol Buffers are a way of encoding structured data in an
+         efficient yet extensible format.  Google uses Protocol Buffers for
+         almost all of its internal RPC protocols and file formats.
+      '';
+
+    license = "mBSD";
```
Comment (jagajaga) [inline pkgs/development/libraries/protobuf/3.0.nix]: Please use `stdenv.lib.licenses.whateverYouNeed`.

File: pkgs/development/libraries/protobuf/3.0.nix [hunk 5]
```diff
@@ -1,12 +1,43 @@
...
+
```
Comment (jagajaga) [inline pkgs/development/libraries/protobuf/3.0.nix]: Please omit empty new lines in `meta`.

File: COPYING (hunk 2 above)
Comment (jagajaga) [inline COPYING]: > Because according my terms of employment say that they own the copyright on anything I produce while I work for them.
Can you do that at home at your free time? :)

File: pkgs/development/libraries/protobuf/3.0.nix [hunk 6]
```diff
@@ -1,12 +1,43 @@
...
+    homepage = https://developers.google.com/protocol-buffers/;
```
Comment (jagajaga) [inline pkgs/development/libraries/protobuf/3.0.nix]: Please set `platforms   = platforms.unix;`.

File: COPYING (hunk 2 above)
Comment (zimbatm) [inline COPYING]: Copyright on this repo is implicitly shared between all the authors (if your country adheres to the [Berne convention](https://en.wikipedia.org/wiki/Berne_Convention)). It's correct to update the COPYING file to reflect that. Our mistake is in not doing it sooner.

Comment (eduarrrd) [inline COPYING]: @jagajaga Thanks for the suggestion but I am doing this at home in my free time. Can you elaborate why the notice is unacceptable to you? Do you believe it would be factually incorrect?

File: pkgs/development/libraries/protobuf/3.0.nix (hunk 3 above)
Comment (eduarrrd) [inline pkgs/development/libraries/protobuf/3.0.nix]: Thanks, done.

File: pkgs/development/libraries/protobuf/3.0.nix (hunk 4 above)
Comment (eduarrrd) [inline pkgs/development/libraries/protobuf/3.0.nix]: Thanks, done.

File: pkgs/development/libraries/protobuf/3.0.nix (hunk 5 above)
Comment (eduarrrd) [inline pkgs/development/libraries/protobuf/3.0.nix]: Thanks, done.

File: pkgs/development/libraries/protobuf/3.0.nix (hunk 6 above)
Comment (eduarrrd) [inline pkgs/development/libraries/protobuf/3.0.nix]: Thanks, done.

File: COPYING (hunk 2 above)
Comment (jagajaga) [inline COPYING]: Google Inc. is not related to NixOS.

Comment (peti) [inline COPYING]: Listing Google Inc. as the only explicitly mentioned copyright holder next to Eelco does feel excessive. I also doubt that it is necessary to do so. Generally speaking, we need a better policy with regard to (c) issues that what we have today, i.e. mentioned Eelco as the only (c) holder doesn't feel quite right either. Anyway, he's certainly the most important contributor to Nixpkgs, so the status quo feels okay'ish to me until we have figured out a better solution.

Comment (eduarrrd) [inline COPYING]: I don't understand this reasoning? In your view, what is this file for? What would have if Eelco quits nixos? Will he be removed from this file?

Comment (eduarrrd) [inline COPYING]: @peti I'm not sure why people didn't add their names either. If you want, I can extract the names of all committers from the git log and add them here so there won't be false exclusivity. How does that sound?

Comment (peti) [inline COPYING]: The fact that Eelco is the only person mentioned in that file is a historical curiosity. He started the project, he created that file, and he's the only one who has ever touched it since (in 6b0b4586945e3531a04561f4fcffd7e5f666513d). IMHO, the contents of that file is inaccurate and meaningless, at best, and we need to improve the way we handle copyright issues.

Anyway, I agree with @jagajaga that adding Google to that license statement seems odd and I'm not in favor of doing so.

Comment (zimbatm) [inline COPYING]: It might be easier to maintain a list of contributors like golang does: https://github.com/golang/go/blob/master/CONTRIBUTORS

Comment (kernc) [inline COPYING]: @eduarrrd Google does _not_ hold the copyright to patches you contribute at your home in your free time.

Comment (peti) [inline COPYING]: > I can extract the names of all committers from the git log and add them here so there won't be false exclusivity. How does that sound?

I think it would be best to do that (a) in a separate PR and (b) only after discussing this topic on the nix-dev mailing list. This is not something that we can decide easily.

Comment (jagajaga) [inline COPYING]: @peti but I think that discussion about exactly this PR is over after @kernc's message?

Comment (silverhook) [inline COPYING]: (I’m quite busy, so it’s a hurried reply, sorry)

@eduarrrd If you created that in your free time, it doesn’t count as [“work for hire” (link: Wikipedia)][wfh] even under US copyright.

To quote [Columbia Law School’s article][cls] on this topic:

> On the other hand, if you’ve written a novel or a computer application in your spare time, and that isn’t your job, you, not your employer, own those rights. That’s important to remember for the many creators who also have day jobs.

In any case you would be entitled to moral rights under the Berne convention, but *if* you assigned the rights to the work you did here to Google (e.g. as work for hire or a separate contract), it might indeed be that Google has to be listed as a copyright holder.

Without seeing the details, the situation you describe here, it indeed sounds like you wrote this in your free time and Google has no right to be listed as copyright holder. The one who has the right to be listed is you, who wrote it.

As @zimbatm noticed, all the authors should have been listed in the copyright headers and the `COPYING` or `AUTHORS` file already before. So that needs to be fixed at some point in any case.


[wfh]: https://en.wikipedia.org/wiki/Work_for_hire
[cls]: http://web.law.columbia.edu/keep-your-copyrights/copyrights/works-made-hire

Comment (eduarrrd) [inline COPYING]: @kernc I'm not what makes you the expert on my terms of employment but the
top comment might clear things up for you:
https://news.ycombinator.com/item?id=1969979

Comment (jagajaga) [inline COPYING]: @eduarrrd send me your code, I will remove that duty from you and commit under my own name :)
Indeed. C'mon. You are person, not company.

Comment (peti) [inline COPYING]: @jagajaga,

> I think that discussion about exactly this PR is over after @kernc's message?

I believe @eduarrrd is the only one qualified to judge whether this patch is his private property or that of his employer, so I'd rather not presume to tell him what to do.

What I *can* say is that I, personally, will not merge a patch that modifies the top-level `COPYING` file without getting explicit consent from a representative part of the NixOS community first. If this means that this patch cannot be merged, then that's unfortunate but it's the way it is. Other collaborators might feel differently about this and they might be willing to merge the patch as-is. If that is so, then more power to them! It's just that I won't.

Comment (jagajaga) [inline COPYING]: I agree with you wholeheartedly, @peti.

Comment (silverhook) [inline COPYING]: @eduarrrd Ah so it’s additional terms of contract. Yes, in US you can probably do that.

It could be argued that you’re not competing with Google with this contribution, but without seeing the contract etc. etc. it’s just guesswork. You’d probably need to get a lawyer, if you wanted to try that.

(also, of course, neither this or my post above is legal advice)

Comment (eduarrrd) [inline COPYING]: #11557 is the followup. IF that one gets merged I will remove the COPYING change from this one. Post to nix-dev to follow regarding the #11557 to follow later today.

**Comment by jagajaga** [comment]
Please don't forget to properly name commit following our guides in `CONTRIBUTING.md`/

File: COPYING (hunk 2 above)
Comment (domenkozar) [inline COPYING]: cc @wkennington who might be interested into this debate.

Comment (vcunat) [inline COPYING]: I'm certainly against adding that line.

Comment (eduarrrd) [inline COPYING]: @vcunat so is everyone ;-). It currently serves a "do not merge" function.
If #11557 merges this line will not be part of this PR.

**Comment by oconnorr** [comment]
@eduarrrd you should be able to continue this PR now.

**Comment by eduarrrd** [comment]
Update for current upstream beta.

PTAL

File: pkgs/development/libraries/gmock/default.nix [hunk 7]
```diff
@@ -1,15 +1,13 @@
-{ stdenv, fetchurl, unzip, cmake}:
-
+{ stdenv, cmake, callPackage }:
+let
+  source = callPackage ./source.nix { };
```
Comment (vcunat) [inline pkgs/development/libraries/gmock/default.nix]: Why not use `fetchzip` instead?

Comment (vcunat) [inline pkgs/development/libraries/gmock/default.nix]: Or maybe `fetchFromGitHub` directly, as I see the URL redirects to github anyway.

Comment (eduarrrd) [inline pkgs/development/libraries/gmock/default.nix]: @vcunat Done (using fetchzip).

Didn't know about fetchzip. Now I do :-). I didn't implement the fetch, I just split out the maintainer's fetching logic into a separate output so I didn't investigate too deeply.

**Comment by eduarrrd** [comment]
Adressed @vcunat 's concern.

Tested. Travis fails due to (seemingly) unrelated merge conflicts.

PTAL.

File: pkgs/development/libraries/gmock/default.nix (hunk 7 above)
Comment (vcunat) [inline pkgs/development/libraries/gmock/default.nix]: That was a partial misunderstanding between us. It's enough to just write
```nix
let
  source = fetchzip { /* ... */ };
```
I don't think there's any benefit in making a `mkDerivation` wrapper around that to only copy the same thing into another path. If you want to override `name`, that can also be passed to `fetchzip` directly.

Comment (eduarrrd) [inline pkgs/development/libraries/gmock/default.nix]: I took this pattern from pkgs/development/libraries/gtest/default.nix . Does that one use it better than I do?
Do you have an example of what you are proposing? I'm not sure I understand your proposal correctly.

Comment (vcunat) [inline pkgs/development/libraries/gmock/default.nix]: The thing is that `fetchurl` only fetches, but `fetchzip` (and derived functions) also unpacks the archive afterwards, all in a single step. So, `gtest` could also use `fetchzip` instead of the whole gtest-src derivation. It might be just because `fetchzip` is relatively new and anyway, its main purpose is to fight non-deterministic archives.

See https://github.com/eduarrrd/nixpkgs/pull/1

**Comment by eduarrrd** [comment]
@vcunat Many thanks for your "example" :-).

**Comment by vcunat** [comment]
I also enabled parallel building, as I didn't like to wait too long.
